File: Linked_List/138.Copy_List_with_Random_Pointer/C++/138.Copy_List_with_Random_Pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node * copyRandomList(Node * head) 
    {
        //====================================//
        // Input type:                        //
        //   - Node * (head list-node)        //
        // Return type:                       //
        //   - Node * (result head list-node) //
        //====================================//
        if (! head) { return NULL; } //Issue, Boundary-case handling

        /*Initialize*/
        ///// Record dictionary, hashmap /////
        unordered_map<Node * , Node * > record_map;

        ///// Result, Record head list-node /////
        Node * res_head, * record_head = recurseTraversal(head, record_map);

        res_head = record_head;


        /*Recursion-based loop traversal with recorded dictionary, hasmap*/
        while (head) //Whole
        {
            ///// Check if the current random pointer existed or not /////
            if ((head->random))
            {
                (record_head->random) = record_map[(head->random)]; //Keep updating, overwriting
            }
            else { ; }

            head = (head->next); //Keep updating, overwriting
            record_head = (record_head->next); //Keep updating, overwriting

        } //Whole

        return res_head;
    }


    Node * recurseTraversal(Node * head, unordered_map<Node * , Node * > & record_map)
    {
        //=======================================================================//
        // Input type:                                                           //
        //   - Node * (head list-node)                                           //
        //   - unordered_map<Node * , Node * > ref. (record dictionary, hashmap) //
        // Return type:                                                          //
        //   - Node * (result head list-node)                                    //
        //=======================================================================// 
        if (! head) { return NULL; } //Issue, Boundary-case handling

        /*Whole process, flow*/
        Node * res_head = new Node((head->val)); //Result list-node head

        (res_head->next) = recurseTraversal((head->next), record_map); //Update, Overwrite

        (res_head->random) = NULL; //Update, Overwrite

        record_map[head] = res_head; //Update, Record

        return res_head;
    }
};

File: Linked_List/138.Copy_List_with_Random_Pointer/Python/138.Copy_List_with_Random_Pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        #==================================#
        # Recursion-based traversal method #
        #==================================#
        if (not head): #Issue/Boundary-case handling
            return None

        ############
        #Initialize
        ##### Record dictionary/hashmap #####
        record_dict = {}

        ##### Result, Record head list-node #####
        res_head = record_head = self.recurseTraversal(head, record_dict)


        #################################################################
        #Recursion-based loop traversal with recorded dictionary/hashmap
        while (head):

            ##### Ceck if the current random pointer existed or not #####
            if (head.random):
                record_head.random = record_dict[head.random] #Keep updating/overwriting
            else:
                pass

            head = head.next #Keep updating/overwriting
            record_head = record_head.next #Keep updating/overwriting

        return res_head


    def recurseTraversal(self, head, record_dict):
        """
        :type head: Node
        :type record_dict: dict
        :rtype: Node
        """
        #====================================#
        # Recursion-based traversal function #
        #====================================#
        if (not head): #Issue/Boundary-case handling
            return None

        ####################
        #Whole process/flow
        res_head = Node(head.val) #Result head list-node

        res_head.next = self.recurseTraversal(head.next, record_dict) #Update/Overwrite

        res_head.random = None #Update/Overwrite

        record_dict[head] = res_head #Update/Record

        return res_head
